src/hooks/useFetch.js

```diff
@@ -0,0 +1,72 @@
+import { useEffect, useState } from 'react';
+
+const fetchUrl = (url, after, count, limit = 100, time = 'year') => {
+  const urlParams = `t=${time}&limit=${limit}${
+    count && after ? `&after=${after}&count${count}` : ''
+  }`;
+  return `${url}/top.json?${urlParams}`;
+};
+
+const makeRequest = async (url) => {
+  const res = await fetch(url);
+
+  if (!res.ok) {
+    throw new Error('Unable to fetch data');
+  }
+
+  const data = await res.json();
+  return data;
+};
+
+const fetchPosts = async (
+  url,
+  posts = [],
+  count = null,
+  after = null,
+  numRequests = 500,
+) => {
+  if (count >= numRequests) {
+    return posts;
+  }
+
+  const currentUrl = fetchUrl(url, after, count);
+
+  const data = await makeRequest(currentUrl);
+
+  const recordsRetrieved = data.data.dist;
+  posts.push(...data.data.children);
+
+  if (recordsRetrieved < 100) {
+    return posts;
+  }
+
+  const currentCount = count + recordsRetrieved;
+  const currentAfter = data.data.after;
+
+  return fetchPosts(url, posts, currentCount, currentAfter);
+};
+
+const useFetch = (url) => {
+  const [response, setResponse] = useState(null);
+  const [loading, setLoading] = useState(true);
+  const [hasError, setHasError] = useState(false);
+
+  useEffect(() => {
+    setLoading(true);
+
+    const fetchData = async () => {
+      try {
+        const data = await fetchPosts(url);
+        setResponse(data);
+        setLoading(false);
+      } catch (error) {
+        setLoading(false);
+        setHasError(true);
```

Comment:
Here is an interesting issue that I never encountered. I mentioned in the posts table that posts can be `null` and that's why some Cypress tests are failing. But obviously you handle errors here. Unfortunately, there's a bug.

To reproduce the issue you can change the domain to some non-existing domain like `https://www.redditdsad.com`. Now the request will fail and you can see the error.

If you like try to find the bug yourself. It's interesting but pretty nasty. So if you like I can also provide you the answer :wink:

Reply:
Not sure if you investigated this bug but you fixed it by using `finally`. If you're interested, here's what went wrong before:

The problem was that `loading` was set to `false` before `hasError` was set to `true`. Usually, React batches state updates, but if they're wrapped in a promise the states are updated separately. [Here's a stackoverflow thread about this](https://stackoverflow.com/questions/53048495/does-react-batch-state-update-functions-when-using-hooks). Never happened to me before :smile: